File: Silver II/1802/yeolyi.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

input.slice(1).forEach(x => {
    const arr = [...x].map(x => x==='0');
    console.log(solve(arr) ? 'YES' : 'NO');
})

function solve(arr) {
    if (arr.length === 1) { return true; }
    let length = arr.length;
    let middle = Math.floor(length/2)
    let left = arr.slice(0, middle);
    let right = arr.slice(middle+1);

    for (let i=0; i<left.length; i++) {
        if (left[i] === right[middle-i-1]) {
            return false
        }
    }

    return solve(left);
}

File: Gold IV/1976/yeolyi.py
N = int(input())
input()
arr = [[x == '1' for x in input().split()] for _ in range(N)]
route = [x-1 for x in map(int, input().split())]


def findParent(parent, x):
    if (parent[x] == x):
        return x
    parent[x] = findParent(parent, parent[x])
    return parent[x]


def union(parent, a, b):
    a = findParent(parent, a)
    b = findParent(parent, b)
    parent[a] = b


parent = [x for x in range(N)]
for i in range(N):
    for j in range(i):
        if arr[i][j]:
            union(parent, i, j)

routeParent = findParent(parent, route[0])

for x in route:
    if findParent(parent, x) != routeParent:
        print('NO')
        break
else:
    print('YES')
